src/plugins/discourse/fetch.js

```diff
@@ -0,0 +1,212 @@
+  latestTopicId(): Promise<TopicId>;
+  // Retrieve the Topic with Posts for a given id. May be null, either because the
+  // topic is hidden from the api user, or it 404s. (Not sure why this happens.)
+  topicWithPosts(id: TopicId): Promise<TopicWithPosts | null>;
+  // Retrieve an individual Post by its id. May be null, e.g. because the post is
+  // hidden from the api user.
+  post(id: PostId): Promise<Post | null>;
+  // Retrieve the latest posts from the server.
+  // Vital so that we can then enumerate and fetch every Post that we haven't
+  // encountered.
+  latestPosts(): Promise<$ReadOnlyArray<Post>>;
+}
+
+export class DiscourseFetcher implements DiscourseInterface {
+  +options: DiscourseFetchOptions;
+  +_fetchImplementation: typeof fetch;
+
+  constructor(
+    options: DiscourseFetchOptions,
+    // fetchImplementation shouldn't be provided by clients, but is convenient for testing.
+    fetchImplementation?: typeof fetch
+  ) {
+    this.options = options;
+    this._fetchImplementation = fetchImplementation || fetch;
+  }
+
+  _fetch(endpoint: string): Promise<Response> {
+    const {serverUrl, apiKey, apiUsername} = this.options;
+    if (!endpoint.startsWith("/")) {
+      throw new Error(`invalid endpoint: ${endpoint}`);
+    }
+    if (!serverUrl.startsWith("http") || serverUrl.endsWith("/")) {
+      throw new Error(`invalid server url: ${serverUrl}`);
+    }
+    const fetchOptions = {
+      method: "GET",
+      headers: {
+        "Api-Key": apiKey,
+        "Api-Username": apiUsername,
+        "Content-Type": "application/json",
+      },
+    };
+    const fullUrl = `${serverUrl}${endpoint}`;
+    return this._fetchImplementation(fullUrl, fetchOptions);
+  }
+
+  async latestTopicId(): Promise<TopicId> {
+    const response = await this._fetch("/latest.json?order=created");
+    if (response.status === 404) {
+      throw new Error(`404 from ${response.url}; maybe bad serverUrl?`);
+    }
+    if (response.status === 403) {
+      const {apiUsername, apiKey} = this.options;
+      throw new Error(
+        `got status 403; bad api username (${apiUsername}) or key (${apiKey})`
+      );
+    }
+    if (response.status !== 200) {
+      throw new Error(`not OK status ${response.status} on ${response.url}`);
+    }
+    const json = await response.json();
+    if (json.topic_list.topics.length === 0) {
+      throw new Error(`no topics! got ${stringify(json)} as latest topics.`);
+    }
+    return json.topic_list.topics[0].id;
+  }
+
+  async latestPosts(): Promise<$ReadOnlyArray<Post>> {
+    const response = await this._fetch("/posts.json");
+    if (response.status === 404) {
+      throw new Error(`404 from ${response.url}; maybe bad serverUrl?`);
+    }
+    if (response.status === 403) {
+      const {apiUsername, apiKey} = this.options;
+      throw new Error(
+        `got status 403; bad api username (${apiUsername}) or key (${apiKey})`
```

Comment:
Again on the security note.

Perhaps encapsulate this logic into a function? It’s duplicated verbatim
across these two functions.